fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt

```diff
@@ -278,45 +283,68 @@ class AwesomeBarView(
         }
 
         if (state.showAllHistorySuggestions) {
-            if (activity.settings().historyMetadataUIFeature) {
-                providersToAdd.add(defaultCombinedHistoryProvider)
-            } else {
-                providersToAdd.add(defaultHistoryStorageProvider)
-            }
+            providersToAdd.add(
+                getHistoryProvider(
+                    filter = getFilterToExcludeSponsoredResults(state),
+                ),
```

Comment:
We talked about this offline, but I wanted to capture that discussion here, so that we can leave some breadcrumbs for our future selves!

Previously—on tip-of-tree Fenix today; before my PR—we have a few cases:

* If `state.showAllHistorySuggestions == true`, we use `defaultCombinedHistoryProvider` or `defaultHistoryProvider`. These actually [don't do any filtering at all](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L120-L140), and omit the `resultsUriFilter` param.
* If `state.showHistorySuggestionsForCurrentEngine == true`, we call `getHistoryProvidersForSearchEngine()`. That method [returns `null`](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L354) if `state.searchEngineSource.searchEngine == null`, or a new provider (combined or not) with a `resultUriFilter` param, and the [same other options](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L358-L365) as the default providers.

To make things even more complicated, the way we handle "`state.searchEngineSource.searchEngine == null`" for history is consistent with how we handle [search term suggestions](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L402), but inconsistent with the way we handle [bookmarks](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L535-L538), [synced tabs](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L471-L474), and [local tabs](https://github.com/mozilla-mobile/firefox-android/blob/ff06d1ef8f05887291213d0ea0d8c25eb999e15a/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#L504-L507). Bookmarks, synced tabs, and local tabs will return a provider _without_ a filter if we're only showing suggestions for the current filter, but `state.searchEngineSource.searchEngine == null`.

And to get even further into the weeds, I think the only case where `searchEngineSource.searchEngine == null` is for [`SearchEngineSource.None`](https://github.com/mozilla-mobile/firefox-android/blob/f2afc4e57598b31fdab25600b575de354a53cc2a/fenix/app/src/main/java/org/mozilla/fenix/search/SearchFragmentStore.kt#L38-L43). IIUC, that's only used (1) in test code, or (2) when the user hasn't selected a search engine, and we don't have a default for their region.

With my PR, things are a little different:

* `getHistoryProvidersForSearchEngine()` is now `getHistoryProvider()`, and behaves like the other `get*Provider()` methods.
*  If `state.showAllHistorySuggestions == true` and `state.showSponsoredSuggestions == false`, `getHistoryProvider()` returns `defaultCombinedHistoryProvider` or `defaultHistoryProvider`.
* If `state.showAllHistorySuggestions == true` and `state.showSponsoredSuggestions == true`, `getHistoryProvider()`returns a (combined or not) history provider, with a filter to exclude sponsored suggestions.
* If `state.showHistorySuggestionsForCurrentEngine == true`, we call `getFilterForCurrentEngineResults()`. If that returns `null`—in the case where `searchEngineSource.searchEngine == null`—we don't call `getHistoryProvider()` at all. Otherwise, we call `getHistoryProvider()` with that filter, which will only include suggestions with the same host as `searchEngineSource.searchEngine.resultsUrl`.

I also changed bookmarks, synced tabs, and local tabs to work like history in the case where `searchEngineSource.searchEngine == null`. I don't think that this is a user-visible change—because I don't think there's any way to select a `None` engine—but it was more for consistency later, so that we aren't wondering, "why do we not show any history for `None`, but all bookmarks and tabs?"